test/test_python_dispatch.py

```diff
@@ -335,6 +336,34 @@ def backward(ctx, grad_output):
 $5 = torch._ops.aten.mul($4, $0)
 $6 = torch._ops.aten.add_($1, $5)''')
 
+    def test_subclass_creation(self):
+        # Make sure these statements runs without error
+        # In particular checking that when internal detach returns
+        # subclasses, these are cleanly overwritten.
+        class Foo(torch.Tensor):
+            pass
+        a = torch.Tensor._make_subclass(Foo, LoggingTensor(torch.rand(2)))
+        b = LoggingTensor(torch.rand(2)).as_subclass(Foo)
+
+        # And in case where we don't know if the user wants this subclass
+        # overwritten, raise a nice error.
+        # The standard LoggingTensor will fail because it is not on the right device
+        with self.assertRaisesRegex(TypeError, "expected.*device=cpu.*device=meta"):
```

Comment:
TBH, the device error is not very good, we'd prefer the other error

Reply:
The problem is that this error is raised by the tensor options check in the legacy constructor before we get to the actual subclass wrapping.
This is true in general for the legacy Tensor constructor:
```python
import torch

a = torch.rand(2, device="gpu")
b = torch.Tensor(a)
```
This will raise this same device error.